### PR TITLE
Stops instance before detaching EBS Volume.

### PR DESCRIPTION
**Pull request by c4milo**
Addresses #2957. Terraform's current behavior is to attempt to detach
the EBS volume before destroying or stopping the associated EC2 Instance.
This behavior will cause the volume to get stuck on busy state and could
possibly damage the data it contains.

The current commit stops the EC2 instance before attempting to detach its
EBS Volume. It may not be the best solution, but it is good enough
until a more generalized approach is implemented.

### CODE REVIEW

File: builtin/providers/aws/resource_aws_volume_attachment.go [hunk 1]
```diff
@@ -153,6 +153,19 @@ func resourceAwsVolumeAttachmentDelete(d *schema.ResourceData, meta interface{})
 	vID := d.Get("volume_id").(string)
 	iID := d.Get("instance_id").(string)
 
+	// Stop instance ID using the Volume to avoid loosing data.
```
Comment (SpencerBrown) [inline builtin/providers/aws/resource_aws_volume_attachment.go]: s/loosing/losing/

Comment (c4milo) [inline builtin/providers/aws/resource_aws_volume_attachment.go]: fixed! thank you!

**Comment by sean-bennett112** [comment]
Any update on whether this can get merged in? It'd be great to get this into 0.6.13, since #2957 is kind of a showstopper for using `aws_volume_attachment` instead of declaring volumes inside an individual instance.

**Comment by rmzi** [comment]
+1 to @sean-bennett112. 

@sean-bennett112: Are you saying you can define the volumes you need as part of the instance and persist them across terraform apply's?

**Comment by sean-bennett112** [comment]
@rmzi Provided the instance isn't destroyed on later apply's, I _believe_ that the volumes should stick around. However don't quote me on that, as our use case for this is slightly different, so I haven't really verified that.

**Comment by rmzi** [comment]
Any word on this?

**Comment by c4milo** [comment]
@radeksimko any word on this one? is there any preference on the approach to solve this issue?

**Comment by radeksimko** [comment]
Hi @c4milo ,
sorry for the silence.

Couple of questions:

1. What/who/when is going to turn the instance back on again? At the moment it looks like it would remain in "stopped" state.
2. How can we communicate this behaviour to the user? i.e. attaching/detaching may cause downtime.

If we're going to allow this behaviour, I believe it should be optional. Some people may prefer uptime over potential dataloss (or more precisely they can avoid dataloss by making sure nothing is reading/writing from that volume).

Dealing with mostly ephemeral machines for some time in AWS, I may not entirely understand the workflow/intentions here. I understand there can be some edge cases where long-lived "pet-like" instances are preferred though - e.g. Github Enterprise.

Would you mind explaining workflow for some common real-life-examples of changing EBS attachments in production-like environment (such environments obviously require both high uptime & no dataloss)?

> This behavior will cause the volume to get stuck on busy state and could
> possibly damage the data it contains.

but the instance is still running and if you had unmounted (or remounted as ro) the volume from within the instance, it would be ok, wouldn't it?

**Comment by c4milo** [comment]
@radeksimko, the use case was destroying an instance with an EBS volume attached, it would never get destroyed because Terraform attempts to first detach the volume and then terminate the instance due to how the dependency graph is normally walked.

> 1. What/who/when is going to turn the instance back on again? At the moment it looks like it would remain in "stopped" state.  
> 2. How can we communicate this behaviour to the user? i.e. attaching/detaching may cause downtime.

Given my use case I wasn't seeing this change would not work for people who needs to detach the volume without stopping or destroying the instance. Bummer.

>but the instance is still running and if you had unmounted (or remounted as ro) the volume from within the instance, it would be ok, wouldn't it?

My use case assumed no manual interaction with the instance when terminating it.

If we had a way of running a provisioner upon destroying an instance, we should be able to unmount the volume and allow automatic termination from Terraform.

**Comment by radeksimko** [comment]
> the use case was destroying an instance with an EBS volume attached

understood. I think it would be ok to stop & start instance before it's destroyed, just to correctly detach the EBS volume - i.e. this approach probably makes sense. 👍 

There currently isn't any easy way to find out whether another resource, like `aws_instance` is going to be destroyed in the same run AFAIK, so we can't easily decide here whether or not to start the instance again.

This change would inherently make destroying any instance with any number of `aws_volume_attachment`'s potentially significantly slower. Also some users may have N `aws_volume_attachment`s attached to the same instance, which would effectively cause the instance to be stopped&started N-times.

I think these are all acceptable tradeoffs though.

> Addressing #386 should resolve the issue.

That will likely _allow_ resolving this issue, but in quite a _hacky_ way.

Your use case however highlights another issue - I haven't looked too deeply into the full lifecycle of EC2 instance, but I'm thinking that attempt of terminating an instance which is just booting may error out. We would likely need **retry logic** there to ensure that `aws_volume_attachment` is only considered as "destroyed" when the instance is in the `running` state again.

Also I assume that stopping the instance may take some time and we'd need to have **retry logic** there as well that would verify the instance is actually stopped and we can carry on detaching the EBS volume.

In addition to all the retry logic we might need to add [mutex](https://github.com/hashicorp/terraform/blob/master/helper/mutexkv/mutexkv.go) to avoid multiple `aws_volume_attachment` resources fighting with each other and trying to stop/start the same instance due to the way Terraform executes CRUD of resources in the graph (i.e. in parallel where possible).

> My use case assumed no manual interaction with the instance when terminating it.

Agreed, destroying should be fully automated, I was more thinking along the lines of swapping volumes with bigger ones etc. which I assume will always require some manual interaction.

Regarding the UX, I would make this optional under `force_detach = false`. What should be default state is another question. Are people more likely going to cry over dataloss or downtime? I'd personally say dataloss, so `false` should be probably default value - i.e. slower destroy of instances with EBS volumes by default.

**Comment by c4milo** [comment]
@radeksimko thanks for the thorough explanation! But, I'm not fully convinced yet. I don't think this PR is the way to go to fix this well. We should try to figure out how to make the EC2 instance to unmount the volume before the volume is detached, doing that will solve both problems: dataloss and downtime.

**Comment by SpencerBrown** [comment]
Between the issues that this and #6050 are trying to fix, using Terraform to attach volumes to EBS instances simply does not work in any useful way. This needs serious attention. I've been watching these issues for several months now, and testing some PRs.

Personally I don't have the time to fix Terraform any more so I am going to seriously look at doing my EC2 instance deployments in other ways (e.g. Ansible). Terraform is OK for managing pretty much everything in AWS that doesn't change much once you deploy it. But for things like EC2 instances that we want to have come and go, not so much.

**Comment by Jonnymcc** [comment]
I have been dealing with this issue by tainting the EC2 instance when I know that it will have problems detaching the volume. I also think this would be an acceptable workaround in most cases. Something I think would be useful is that when an EBS volume attachment is planned to be destroyed while it is attached to an instance, `terraform apply` will warn that the EC2 instance resource should be tainted before continuing. In situations where there are multiple instance resources that need to be tainted, it would be nice if along with the warning `terraform apply` also printed the taint commands that need to be run. Or instead of printing the commands, prompt the user saying that these EC2 instances need to be tainted first. Of course if the EC2 instance is not ephemeral and you care about it being terminated then I suppose this is not a great solution.

**Pull request closed without merging**